pkg/services/sqldb/mssql.go

```diff
@@ -7,25 +7,41 @@ import (
 )
 
 type module struct {
-	serviceManager *serviceManager
+	allInOneServiceManager   *allInOneServiceManager
+	serverOnlyServiceManager *serverOnlyServiceManager
 }
 
 type serviceManager struct {
 	armDeployer  arm.Deployer
 	mssqlManager mssql.Manager
 }
 
+type allInOneServiceManager struct {
+	serviceManager
```

Comment:
Does this result in us having to further qualify the access path to get down to an `armDeployer` or `mssqlManager`? i.e. Do we have to do `s. serviceManager. armDeployer` where we could previously do just `s.armDeployer`?

Reply:
No. I refactored in the sqldb/provision.go like so:

In provision.go, now we have two methods that take care of building up proper parameters and determining which arm template to use. these are receivers for the two new service managers. 
>func (s *serverOnlyServiceManager) deployARMTemplate(....)
>func (a *allInOneServiceManager) deployARMTemplate(

I refactored the old logic about actually calling the arm deployer like so:
>func (s *serviceManager) deployTemplate(......)

Through embedding, this applies to both and it's getting the old service manager. It also gets the arm template + the params as method parameters.  It's fairly generic now and allows a little more DRY in there and can be used by these two + our third (db only) and any other.

Reply:
Ah... wasn't sure what the benefit to this was, at first, but now I see that there are spots where you've defined functions required by the `ServiceManager` interface with `*serviceManager` (the "base class" if this were OO) as the receiver. `GetDeprovisioner(...)` is one such example...

I'm not sure how I feel about this, because anywhere we do that, we're going to require that the types those functions operate on are common to the multiple `ServiceManager` implementations in question-- and that might not be ideal, because the service-specific details are still going to vary from one `Service` to the next...

For example, I see that this context type is common to both serviceManagers:

```
type mssqlProvisioningContext struct {
	ARMDeploymentName          string `json:"armDeployment"`
	ServerName                 string `json:"server"`
	AdministratorLogin         string `json:"administratorLogin"`
	AdministratorLoginPassword string `json:"administratorLoginPassword"`
	DatabaseName               string `json:"database"`
	FullyQualifiedDomainName   string `json:"fullyQualifiedDomainName"`
}
```

While we _can_ make that work, it's a little awkward.

What do you think about separating _all_ the types and functions so that all are unique to a given `ServiceManager` implementation?

Eventually, where I think we'll land is that _after_ we implement the "database only" service, we'll have a context type for server only and another context type for database only and there won't be much, if any overlap between the two. _Then_ we can maybe make the "all in one" context be a composition of the other two.

Reply:
Yeah, I agree. For the provisoner case, they were mostly similar (except that database), but like I mentioned in a comment on the Bind method, I think more explicit is better in the end and once we implement that database only step, it will for sure be a composition of the two. I’ll decompose more.